crucible/from-file-json-manipulation-api/run.sh

```diff
@@ -0,0 +1,200 @@
+    fi
+}
+
+# change an existing run-params parameter value
+function override_run_param() {
+    local param=$1; shift
+    local value=$1; shift
+
+    do_jq --arg param "${param}" --arg value "${value}" '."run-params"[$param] = $value'
+}
+
+# add a new parameter to the run-params object
+function add_run_param() {
+    local param=$1; shift
+    local value=$1; shift
+
+    do_jq --arg param "${param}" --arg value "${value}" '."run-params" += { ($param): $value }'
+}
+
+# disable tools by emptying the array
+function no_tools() {
+    do_jq '."tool-params" = []'
+}
+
+# add a tag to the tags object
+function add_tag() {
+    local tag=$1; shift
+    local value=$1; shift
+
+    do_jq --arg tag "${tag}" --arg value "${value}" '."tags" += { ($tag): $value }'
+}
+
+# insert a new endpoint object to the endpoints array at index 0
+function add_endpoint_obj() {
+    do_jq '."endpoints" |= [{}] + .'
+}
+
+# add a new property to the endpoint object at endpoints array index 0
+function add_endpoint_obj_property() {
+    local property=$1; shift
+    local value=$1; shift
+
+    do_jq --arg property "${property}" --arg value "${value}" '."endpoints"[0] += { ($property): $value }'
+}
+
+# add an empty config array property to the endpoint object at
+# endpoints array index 0
+function add_endpoint_cfg() {
+    do_jq '."endpoints"[0] += { "config": [] }'
+}
+
+# insert a config object to the config array at index 0 for the
+# endpoint object at endpoints array index 0
+function add_endpoint_cfg_obj() {
+    do_jq '."endpoints"[0]."config" |= [{ "settings": {} }] + .'
+}
+
+# create a targets property with the default value to the config
+# object at config array index 0 for the endpoint object at endpoints
+# array index 0
+function add_endpoint_cfg_obj_default_target() {
+    do_jq '."endpoints"[0]."config"[0] += { "targets": "default" }'
+}
+
+# create a an empty targets array property to the config object at
+# config array index 0 for the endpoint object at endpoints array
+# index 0
+function add_endpoint_cfg_obj_target_obj() {
+    do_jq '."endpoints"[0]."config"[0] += { "targets": [] }'
+}
+
+# insert a target object at targets array index 0 to the config object
+# at config array index 0 for the endpoint object at endpoints array
+# index 0
+function add_endpoint_cfg_obj_target_obj_entry() {
+    local role=$1; shift
+    local ids=$1; shift
+
+    do_jq --arg role "${role}" --arg ids "${ids}" '."endpoints"[0]."config"[0]."targets" |= [{ "role": $role, "ids": $ids }] + .'
+}
+
+# add a settings property to the settings object of the config object
+# at config array index 0 for the endpoint object at endpoints array
+# index 0
+function add_endpoint_cfg_obj_setting() {
+    local setting=$1; shift
+    local value=$1; shift
+
+    do_jq --arg setting "${setting}" --arg value "${value}" '."endpoints"[0]."config"[0]."settings" += { ($setting): $value }'
+}
+
+#################################################################################
```

Comment:
suggestions, not a hard requirement... 
functions above: this is the code, the api. Keep it in the main script. BTW, I don't like the term "run.sh"for all of it. I'd rename the api to run-file-api/update-template ...

and below this line: data replacements.... this is "per benchmark", so this is in the run.sh, or a prepare-run-file...

summary of suggestions:
1) api jq function in a separate file under run-file-api/update-template
2) jq calls in run.sh or prepare-run-file (latter implies in running prepare-run-file and then crucible run --from-file...)
3) data replacements abc=123 in a separated file that can be sourced from prepare-run-file